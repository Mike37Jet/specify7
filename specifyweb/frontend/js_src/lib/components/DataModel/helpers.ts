import { f } from '../../utils/functools';
import type { RA } from '../../utils/types';
import { defined } from '../../utils/types';
import { isTreeResource } from '../InitialContext/treeRanks';
import type {
  AnySchema,
  AnyTree,
  SerializedModel,
  SerializedResource,
} from './helperTypes';
import type { SpecifyResource } from './legacyTypes';
import { parseResourceUrl, resourceToJson } from './resource';
import { strictGetModel } from './schema';
import type { Tables } from './types';
import { addMissingFields } from './addMissingFields';

/** Like resource.toJSON(), but keys are converted to camel case */
export const serializeResource = <SCHEMA extends AnySchema>(
  resource: SerializedModel<SCHEMA> | SpecifyResource<SCHEMA>
): SerializedResource<SCHEMA> =>
  serializeModel<SCHEMA>(
    typeof resource.toJSON === 'function'
      ? resourceToJson(resource as SpecifyResource<SCHEMA>)
      : (resource as SerializedModel<SCHEMA>),
    (resource as SpecifyResource<SCHEMA>)?.specifyModel?.name
  );

export const specialFields = new Set([
  'id',
  'resource_uri',
  'recordset_info',
  '_tableName',
]);

/** Recursive helper for serializeResource */
function serializeModel<SCHEMA extends AnySchema>(
  resource: SerializedModel<SCHEMA>,
  tableName?: keyof Tables
): SerializedResource<SCHEMA> {
  const model = strictGetModel(
    defined(
      (tableName as SCHEMA['tableName']) ??
        resource._tableName ??
        parseResourceUrl((resource.resource_uri as string) ?? '')?.[0],
      `Unable to serialize resource because table name is unknown.` +
        (process.env.NODE_ENV === 'test'
          ? `\nMake sure your test file calls requireContext();`
          : '')
    )
  );
  const fields = [...model.fields.map(({ name }) => name), model.idField.name];

  return addMissingFields(
    model.name,
    Object.fromEntries(
      Object.entries(resource).map(([lowercaseFieldName, value]) => {
        let camelFieldName = fields.find(
          (fieldName) => fieldName.toLowerCase() === lowercaseFieldName
        );
        if (camelFieldName === undefined) {
          camelFieldName = lowercaseFieldName;
          if (!specialFields.has(lowercaseFieldName))
            console.warn(
              `Trying to serialize unknown field ${lowercaseFieldName} for table ${model.name}`,
              resource
            );
        }
        if (
          typeof value === 'object' &&
          value !== null &&
          !specialFields.has(camelFieldName)
        ) {
          const field = model.getField(lowercaseFieldName);
          const tableName =
            field === undefined || !field.isRelationship
              ? undefined
              : field.relatedModel.name;
          return [
            camelFieldName,
            Array.isArray(value)
              ? value.map((value) =>
                  serializeModel(
                    value as unknown as SerializedModel<SCHEMA>,
                    tableName
                  )
                )
              : serializeModel(value as SerializedModel<AnySchema>, tableName),
          ];
        } else return [camelFieldName, value];
      })
    )
  ) as SerializedResource<SCHEMA>;
}

export const isResourceOfType = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
  // @ts-expect-error
): resource is SpecifyResource<Tables[TABLE_NAME]> =>
  resource.specifyModel.name === tableName;

export const toTable = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  resource.specifyModel.name === tableName ? resource : undefined;

export const toResource = <TABLE_NAME extends keyof Tables>(
  resource: SerializedResource<AnySchema>,
  tableName: TABLE_NAME
): SerializedResource<Tables[TABLE_NAME]> | undefined =>
  resource._tableName === tableName
    ? (resource as SerializedResource<Tables[TABLE_NAME]>)
    : undefined;

export const toTreeTable = (
  resource: SpecifyResource<AnySchema>
): SpecifyResource<AnyTree> | undefined =>
  isTreeResource(resource) ? resource : undefined;

export const toTables = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableNames: RA<TABLE_NAME>
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  f.includes(tableNames, resource.specifyModel.name) ? resource : undefined;
