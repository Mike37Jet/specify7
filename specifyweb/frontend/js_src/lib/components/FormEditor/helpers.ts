import type { RR } from '../../utils/types';
import type { Tables } from '../DataModel/types';

export const tablesWithBusRulesIn6 = new Set([
  'AccessionAgentBusRules',
  'AccessionAuthorizationBusRules',
  'AccessionBusRules',
  'AddressBusRules',
  'AgentBusRules',
  'AppraisalBusRules',
  'AttachmentBusRules',
  'AttachmentOwnerBaseBusRules',
  'AttachmentReferenceBaseBusRules',
  'AuthorBusRules',
  'BaseTreeBusRules',
  'BioStratBusRules',
  'BorrowBusRules',
  'CatalogNumberingSchemeBusRules',
  'CatAutoNumberingSchemeBusRules',
  'CollectingEventAttributeBusRules',
  'CollectingEventAuthorizationBusRules',
  'CollectingEventBusRules',
  'CollectingTripAuthorizationBusRules',
  'CollectingTripBusRules',
  'CollectionBusRules',
  'CollectionObjectBusRules',
  'CollectionSetupBusRules',
  'CollectorBusRules',
  'ConservDescriptionBusRules',
  'ConservEventBusRules',
  'ContainerBusRules',
  'DataTypeBusRules',
  'DeaccessionAgentBusRules',
  'DeaccessionBusRules',
  'DeterminationBusRules',
  'DeterminerBusRules',
  'DisciplineBusRules',
  'DisposalAgentBusRules',
  'DisposalBusRules',
  'DisposalPreparationBusRules',
  'DivisionBusRules',
  'DNASequenceBusRules',
  'DNASequencingRunBusRules',
  'ExchangeInBusRules',
  'ExchangeOutBusRules',
  'ExchangeOutPrepBusRules',
  'ExtractorBusRules',
  'FieldNotebookBusRules',
  'FieldNotebookPageBusRules',
  'FieldNotebookPageSetBusRules',
  'FundingAgentBusRules',
  'GeoCoordDetailBusRules',
  'GeographyBusRules',
  'GeologicTimePeriodBusRules',
  'GiftBusRules',
  'GiftPreparationBusRules',
  'GroupPersonBusRules',
  'InfoRequestBusRules',
  'InstitutionBusRules',
  'JournalBusRules',
  'LithoStratBusRules',
  'LoanBusRules',
  'LoanGiftShipmentBusRules',
  'LoanPreparationBusRules',
  'LoanReturnPreparationBusRules',
  'LocalityBusRules',
  'LocalityDetailBusRules',
  'PaleoContextBusRules',
  'PcrPersonBusRules',
  'PermitBusRules',
  'PickListBusRules',
  'PreparationBusRules',
  'PrepTypeBusRules',
  'ReferenceWorkBusRules',
  'RepositoryAgreementBusRules',
  'ShipmentBusRules',
  'SpecifyUserBusRules',
  'StorageBusRules',
  'TableSearchResults',
  'TaxonBusRules',
  'TreatmentEventBusRules',
  'TreeableSearchQueryBuilder',
  'TreeDefBusRules',
]);

/**
 * Most of the time business rules class name can be inferred from table name.
 * Exceptions:
 */
export const businessRulesOverride: Partial<RR<keyof Tables, string>> = {
  Shipment: 'LoanGiftShipment',
  CollectionObjectAttachment: 'Attachment',
};

export const getBusinessRuleClassFromTable = (
  tableName: keyof Tables
): string =>
  businessRulesOverride[tableName] === undefined
    ? tablesWithBusRulesIn6.has(`${tableName}BusRules`)
      ? `edu.ku.brc.specify.datamodel.busrules.${tableName}BusRules`
      : ''
    : `edu.ku.brc.specify.datamodel.busrules.${businessRulesOverride[
        tableName
      ]!}BusRules`;
