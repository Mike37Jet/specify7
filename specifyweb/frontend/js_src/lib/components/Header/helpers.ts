export function formatDate(date: Date) {
  const year = date.getFullYear();
  const month = date.getMonth().toString();
  const day = date.getDate().toString();
  const hours = date.getHours().toString();
  const minutes = date.getMinutes().toString();
  const seconds = date.getSeconds().toString();

  return `${year}-${month}-${day} ${hours}:${minutes}:${seconds}`;
}
