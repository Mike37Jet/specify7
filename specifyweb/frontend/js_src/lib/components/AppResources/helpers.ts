import type { AppResourcesConformation } from './Aside';
import type { AppResources, AppResourcesTree } from './hooks';
import type { SpAppResource, SpViewSetObj } from '../DataModel/types';
import { f } from '../../utils/functools';
import type { SpecifyModel } from '../DataModel/specifyModel';
import type { RA } from '../../utils/types';
import { SerializedResource } from '../DataModel/helperTypes';

export const getAppResource = (
  resources: AppResources,
  model: SpecifyModel<SpAppResource | SpViewSetObj>,
  resourceId: number | undefined
): SerializedResource<SpAppResource | SpViewSetObj> | false | undefined =>
  typeof resourceId === 'number'
    ? (
        (model.name === 'SpAppResource'
          ? resources.appResources
          : resources.viewSets) as RA<SerializedResource<SpAppResource>>
      ).find(({ id }) => id === resourceId) ?? false
    : undefined;

export type AppResourceMode = 'appResources' | 'viewSets';

export const getAppResourceMode = (
  resource: SerializedResource<SpAppResource | SpViewSetObj>
): AppResourceMode =>
  resource._tableName === 'SpAppResource' ? 'appResources' : 'viewSets';

export const getAppResourceCount = (
  resourcesTree: AppResourcesTree[number]
): number =>
  resourcesTree.appResources.length +
  resourcesTree.viewSets.length +
  f.sum(resourcesTree.subCategories.map(getAppResourceCount));

export const buildAppResourceConformation = (
  resourcesTree: AppResourcesTree
): RA<AppResourcesConformation> =>
  resourcesTree
    .filter(
      ({ subCategories, appResources, viewSets }) =>
        subCategories.length > 0 ||
        appResources.length > 0 ||
        viewSets.length > 0
    )
    .map(({ key, subCategories }) => ({
      key,
      children: buildAppResourceConformation(subCategories),
    }));
