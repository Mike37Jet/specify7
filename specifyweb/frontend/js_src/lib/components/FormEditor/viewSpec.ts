import type { LocalizedString } from 'typesafe-i18n';

import { f } from '../../utils/functools';
import { tables } from '../DataModel/tables';
import { pipe } from '../Syncer';
import { syncers } from '../Syncer/syncers';
import type { SimpleXmlNode } from '../Syncer/xmlToJson';
import { createSimpleXmlNode } from '../Syncer/xmlToJson';
import { createXmlSpec } from '../Syncer/xmlUtils';

const formDefinitionSpec = f.store(() =>
  createXmlSpec({
    columnDefinitions: pipe(
      syncers.xmlChildren('columnDef'),
      syncers.map(syncers.object(columnDefinitionSpec()))
    ),
    rowDefinition: pipe(
      syncers.xmlChild('rowDef', 'optional'),
      syncers.maybe(syncers.object(rowSizeDefinitionSpec()))
    ),
    rows: pipe(
      syncers.xmlChild('rows'),
      syncers.default<SimpleXmlNode>(() => createSimpleXmlNode('std')),
      syncers.xmlChildren('row'),
      syncers.map(
        pipe(
          syncers.xmlChildren('cell'),
          syncers.map(syncers.object(cellSpec()))
        )
      )
    ),
  })
);

const columnDefinitionSpec = f.store(() =>
  createXmlSpec({
    os: syncers.xmlAttribute('os', 'skip'),
    definition: syncers.xmlContent,
  })
);

const rowSizeDefinitionSpec = f.store(() =>
  createXmlSpec({
    auto: pipe(
      syncers.xmlAttribute('auto', 'skip'),
      syncers.default<LocalizedString>(''),
      syncers.toBoolean
    ),
    cell: syncers.xmlAttribute('cell', 'skip'),
    sep: syncers.xmlAttribute('cell', 'skip'),
    definition: syncers.xmlContent,
  })
);

/**
 * Build a list of tables for which the "formTable" display type should be
 * enabled. This list is not a perfect optimization of what tables have a
 * "formTable" display option in sp6 out of the box, but it's good enough
 */
const tablesWithFormTable = f.store(() =>
  Object.values(tables).filter(
    (table) => !table.isHidden && !table.overrides.isHidden && !table.isSystem
  )
);

const cellSpec = f.store(() =>
  createXmlSpec({
    id: syncers.xmlAttribute('id', 'skip'),
    name: syncers.xmlAttribute('name', 'skip'),
    type: syncers.xmlAttribute('type', 'required'),
    // Make cell occupy more than one column
    colSpan: pipe(
      syncers.xmlAttribute('colSpan', 'skip'),
      syncers.maybe(syncers.toDecimal),
      syncers.default<number>(1)
    ),
    cols: pipe(
      syncers.xmlAttribute('cols', 'skip'),
      syncers.maybe(syncers.toDecimal)
    ),
    rows: pipe(
      syncers.xmlAttribute('colSpan', 'skip'),
      syncers.maybe(syncers.toDecimal)
    ),
    /*
     * In sp6, if false, might make the field invisible. Not used by sp7
     * as it is not implemented consistently by sp6
     */
    legacyVisible: pipe(
      syncers.xmlAttribute('visible', 'skip'),
      syncers.maybe(syncers.toBoolean)
    ),
    // When invisible, field is rendered with visibility="hidden"
    invisible: pipe(
      syncers.xmlAttribute('visible', 'skip'),
      syncers.maybe(syncers.toBoolean),
      syncers.default<boolean>(false)
    ),
    // In sp6, if true, disconnects the field from the database
    legacyIgnore: pipe(
      syncers.xmlAttribute('ignore', 'skip'),
      syncers.maybe(syncers.toBoolean)
    ),
  })
);

export const exportsForTests = {
  tablesWithFormTable,
};
