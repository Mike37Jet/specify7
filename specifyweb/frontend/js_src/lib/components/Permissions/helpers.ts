import { getCache } from '../../utils/cache';
import { f } from '../../utils/functools';
import type { RA } from '../../utils/types';
import { defined, filterArray } from '../../utils/types';
import type { AnyTree } from '../DataModel/helperTypes';
import { schema } from '../DataModel/schema';
import type { Tables } from '../DataModel/types';
import { toolDefinitions } from '../Security/registry';
import { tableNameToResourceName } from '../Security/utils';
import type { tableActions } from './definitions';
import {
  getDerivedPermissions,
  getOperationPermissions,
  getTablePermissions,
} from './index';

const isReadOnly = getCache('forms', 'readOnlyMode') === true;

// REFACTOR: use <ProtectedTable> and etc in favor of this function
/**
 * Security errors are logged so that admins can see why a particular UI
 * component is disabled or missing
 */
export function hasTablePermission(
  tableName: keyof Tables,
  action: typeof tableActions[number],
  collectionId = schema.domainLevelIds.collection
): boolean {
  if (isReadOnly && action !== 'read') return false;
  if (
    defined(getTablePermissions())[collectionId][
      tableNameToResourceName(tableName)
    ][action]
  )
    return true;
  console.log(`No permission to ${action} ${tableName}`);
  return false;
}

export const hasPermission = <
  RESOURCE extends keyof ReturnType<typeof getOperationPermissions>[number]
>(
  resource: RESOURCE,
  action: keyof ReturnType<typeof getOperationPermissions>[number][RESOURCE],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  defined(getOperationPermissions())[collectionId][resource][action]
    ? true
    : f.log(`No permission to ${action.toString()} ${resource}`) ?? false;

export const hasToolPermission = (
  tool: keyof ReturnType<typeof toolDefinitions>,
  action: typeof tableActions[number],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  (toolDefinitions()[tool].tables as RA<keyof Tables>).every((tableName) =>
    hasTablePermission(tableName, action, collectionId)
  );

export const hasTreeAccess = (
  treeName: AnyTree['tableName'],
  action: typeof tableActions[number],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  hasTablePermission(treeName, action, collectionId) &&
  hasTablePermission(`${treeName}TreeDef`, action, collectionId) &&
  hasTablePermission(`${treeName}TreeDefItem`, action, collectionId);

export const hasDerivedPermission = <
  RESOURCE extends keyof ReturnType<typeof getDerivedPermissions>[number]
>(
  resource: RESOURCE,
  action: keyof ReturnType<typeof getDerivedPermissions>[number][RESOURCE],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  defined(getDerivedPermissions())[collectionId][resource][action]
    ? true
    : f.log(`No permission to ${action.toString()} ${resource}`) ?? false;

/** Check if user has a given permission for each table in a mapping path */
export const hasPathPermission = (
  baseTableName: keyof Tables,
  mappingPath: RA<string>,
  action: typeof tableActions[number],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  mappingPathToTableNames(baseTableName, mappingPath, true).every((tableName) =>
    hasTablePermission(tableName, action, collectionId)
  );

export const mappingPathToTableNames = (
  baseTableName: keyof Tables,
  mappingPath: RA<string>,
  ignoreBaseTable = false
): RA<keyof Tables> =>
  f.unique(
    filterArray(
      mappingPath.flatMap((_, index) => {
        if (index === 0 && ignoreBaseTable) return undefined;
        const field = schema.models[baseTableName].getField(
          mappingPath.slice(index).join('.')
        );
        return typeof field === 'object'
          ? [
              field.model.name,
              field.isRelationship ? field.relatedModel.name : undefined,
            ]
          : undefined;
      })
    )
  );
