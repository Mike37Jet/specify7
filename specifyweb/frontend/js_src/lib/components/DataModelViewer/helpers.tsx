import React from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import { schemaText } from '../../localization/schema';
import { syncFieldFormat } from '../../utils/fieldFormat';
import { f } from '../../utils/functools';
import { resolveParser } from '../../utils/parser/definitions';
import { ensure, IR, RA, RR } from '../../utils/types';
import { formatNumber } from '../Atoms/Internationalization';
import { getField } from '../DataModel/helpers';
import { schema } from '../DataModel/schema';
import { TableIcon } from '../Molecules/TableIcon';

export type Value =
  | number
  | string
  | readonly [number | string | undefined, JSX.Element]
  | undefined;
export type Row<SHAPE extends IR<Value>> = SHAPE;

const parser = f.store(() =>
  resolveParser(
    {},
    {
      type: 'java.lang.Boolean',
    }
  )
);

export const booleanFormatter = (value: boolean): string =>
  syncFieldFormat(undefined, parser(), value);

export const tableColumns = f.store(
  () =>
    ({
      name: getField(schema.models.SpLocaleContainer, 'name').label,
      label: schemaText.fieldLabel(),
      isSystem: getField(schema.models.SpLocaleContainer, 'isSystem').label,
      isHidden: getField(schema.models.SpLocaleContainer, 'isHidden').label,
      tableId: schemaText.tableId(),
      fieldCount: schemaText.fieldCount(),
      relationshipCount: schemaText.relationshipCount(),
    } as const)
);

export const getTables = () =>
  ensure<RA<Row<RR<keyof ReturnType<typeof tableColumns>, Value>>>>()(
    Object.values(schema.models).map(
      (model) =>
        ({
          name: [
            model.name.toLowerCase(),
            <>
              <TableIcon label={false} name={model.name} />
              {model.name}
            </>,
          ],
          label: model.label,
          isSystem: booleanFormatter(model.isSystem),
          isHidden: booleanFormatter(model.isHidden),
          tableId: [
            model.tableId,
            <span className="flex w-full justify-end tabular-nums">
              {model.tableId}
            </span>,
          ],
          fieldCount: [
            model.fields.length,
            <span className="flex w-full justify-end tabular-nums">
              {formatNumber(model.fields.length)}
            </span>,
          ],
          relationshipCount: [
            model.relationships.length,
            <span className="flex w-full justify-end tabular-nums">
              {formatNumber(model.relationships.length)}
            </span>,
          ],
        } as const)
    )
  );

export function DataModelRedirect(): null {
  const { tableName = '' } = useParams();
  const navigate = useNavigate();
  React.useEffect(() => {
    navigate(`/specify/data-model/#${tableName}`, { replace: true });
  }, []);
  return null;
}
