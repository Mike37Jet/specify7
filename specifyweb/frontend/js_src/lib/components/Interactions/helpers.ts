import { ajax } from '../../utils/ajax';
import { formData } from '../../utils/ajax/helpers';
import type { RA } from '../../utils/types';

export type Preparations = RA<{
  readonly catalogNumber: string;
  readonly taxon: string;
  readonly preparationId: number;
  readonly prepType: string;
  readonly countAmount: number;
  readonly loaned: number;
  readonly gifted: number;
  readonly exchanged: number;
  readonly available: number;
}>;

export type PreparationRow = readonly [
  string,
  string,
  number,
  string,
  number,
  string | null,
  string | null,
  string | null,
  string
];
export const getPrepsAvailableForLoanRs = async (recordSetId: number) =>
  ajax<RA<PreparationRow>>(
    `/interactions/preparations_available_rs/${recordSetId}/`,
    {
      headers: { Accept: 'application/json' },
    }
  ).then(({ data }) => data);

export const getPrepsAvailableForLoanCoIds = async (
  idField: string,
  collectionObjectIds: RA<string>
) =>
  ajax<RA<PreparationRow>>('/interactions/preparations_available_ids/', {
    method: 'POST',
    headers: { Accept: 'application/json' },
    body: formData({
      id_fld: idField,
      co_ids: collectionObjectIds,
    }),
  }).then(({ data }) => data);

export const getInteractionsForPrepId = async (prepId: number) =>
  ajax<RA<readonly [number, string | null, string | null, string | null]>>(
    '/interactions/prep_interactions/',
    {
      method: 'POST',
      headers: { Accept: 'application/json' },
      body: formData({ prepIds: prepId }),
    }
  ).then(({ data }) => data[0]);

export const getPrepAvailability = async (
  prepId: number,
  iPrepId: number | undefined,
  iPrepName: string
) =>
  ajax(
    `/interactions/prep_availability/${prepId}/${
      typeof iPrepId === 'number' ? `${iPrepId}/${iPrepName}/` : ''
    }`,
    { headers: { Accept: 'application/json' } }
  ).then(({ data }) => data);
