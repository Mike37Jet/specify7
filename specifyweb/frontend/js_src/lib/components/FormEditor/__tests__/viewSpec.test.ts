import { requireContext } from '../../../tests/helpers';
import { exportsForTests, parseSpecifyProperties } from '../viewSpec';

requireContext();

const { tablesWithFormTable, buildSpecifyProperties } = exportsForTests;

test('Tables with form tables computed correctly', () =>
  expect(tablesWithFormTable()).toMatchSnapshot());

const cases = {
  '': {},
  'name=Agent;title=Catalog Agent': {
    name: 'Agent',
    title: 'Catalog Agent',
  },
  'name=PartialDateUI;df=catalogedDate;tp=catalogedDatePrecision': {
    name: 'PartialDateUI',
    df: 'catalogedDate',
    tp: 'catalogedDatePrecision',
  },
  'name=CollectingEvent;clonebtn=true': {
    name: 'CollectingEvent',
    clonebtn: 'true',
  },
  'align=left;': {
    align: 'left',
  },
  'align=right;fg=0,190,0': {
    align: 'right',
    fg: '0,190,0',
  },
  'name=LocalityGeoRef;title=Geo Ref;geoid=geography;locid=localityName;llid=5':
    {
      name: 'LocalityGeoRef',
      title: 'Geo Ref',
      geoid: 'geography',
      locid: 'localityName',
      llid: '5',
    },
  'a=b=c;d=e=f%3B;': {
    a: 'b=c',
    d: 'e=f;',
  },
  'a=;b': {
    a: '',
    b: '',
  },
};

describe('parseSpecifyProperties', () =>
  void Object.entries(cases).forEach(([input, output]) =>
    test(`parses ${input}`, () =>
      expect(parseSpecifyProperties(input)).toEqual(output))
  ));

describe('buildSpecifyProperties', () =>
  void Object.entries(cases).forEach(([output, input]) =>
    test(`parses ${output}`, () =>
      expect(buildSpecifyProperties(input)).toEqual(output))
  ));
