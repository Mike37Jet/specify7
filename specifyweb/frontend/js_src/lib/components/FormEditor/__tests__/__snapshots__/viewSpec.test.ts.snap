// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can edit form definition 1`] = `
"<?xml version='1.0' encoding='UTF-8'?>
<viewdef>
  <desc>The Accession form.</desc>
  <enableRules/>
  <columnDef>100px,2px,175px,5px,140px,2px,161px,5px,75px,2px,160px,0px,15px,p:g</columnDef>
  <columnDef>100px,2px,175px,5px,140px,2px,161px,5px,75px,2px,160px,0px,15px,p:g</columnDef>
  <columnDef os='lnx'>115px,2px,185px,5px,160px,2px,180px,5px,105px,2px,156px,0px,15px,p:g</columnDef>
  <columnDef os='mac'>130px,2px,205px,5px,180px,2px,200px,5px,125px,2px,176px,0px,15px,p:g</columnDef>
  <columnDef os='exp'>p,2px,p:g,5px:g,p,2px,p:g,5px:g,p,2px,p:g,0px,p,p:g</columnDef>
  <rowDef auto='true' cell='p' sep='1px'/>
  <rows>
    <row>
      <cell
        type='field'
        id='1'
        name='accessionNumber'
        uitype='formattedtext'
        initialize=''
      />
      <cell type='label' labelfor='2' initialize=''/>
      <cell type='field' id='2' name='status' uitype='combobox' initialize=''/>
      <cell type='label' labelfor='3' initialize=''/>
      <cell type='field' id='3' name='type' uitype='combobox' initialize=''/>
    </row>
    <row>
      <cell
        type='field'
        id='1'
        name='accessionNumber'
        uitype='formattedtext'
        initialize=''
      />
      <cell type='label' labelfor='2' initialize=''/>
      <cell type='field' id='2' name='status' uitype='combobox' initialize=''/>
      <cell type='label' labelfor='3' initialize=''/>
      <cell type='field' id='3' name='type' uitype='combobox' initialize=''/>
      <cell type='label' labelfor='2' label='New Label' initialize=''/>
    </row>
    <row>
      <cell
        type='field'
        id='1'
        name='accessionNumber'
        uitype='formattedtext'
        initialize=''
      />
      <cell type='label' labelfor='2' initialize=''/>
      <cell type='field' id='2' name='status' uitype='combobox' initialize=''/>
      <cell type='label' labelfor='3' initialize=''/>
      <cell type='field' id='3' name='type' uitype='combobox' initialize=''/>
    </row>
    <row>
      <cell type='label' labelfor='12' initialize=''/>
      <cell
        type='field'
        id='12'
        name='dateAccessioned'
        uitype='formattedtext'
        uifieldformatter='Date'
        default='today'
        initialize=''
      />
      <cell type='label' labelfor='6' initialize=''/>
      <cell
        type='field'
        id='6'
        name='dateReceived'
        uitype='formattedtext'
        uifieldformatter='Date'
        default='today'
        initialize=''
      />
    </row>
  </rows>
</viewdef>"
`;

exports[`Tables with form tables computed correctly 1`] = `
[
  "[table Accession]",
  "[table AccessionAgent]",
  "[table AccessionAuthorization]",
  "[table AccessionCitation]",
  "[table Address]",
  "[table AddressOfRecord]",
  "[table Agent]",
  "[table AgentGeography]",
  "[table AgentIdentifier]",
  "[table AgentSpecialty]",
  "[table Appraisal]",
  "[table Author]",
  "[table Borrow]",
  "[table BorrowAgent]",
  "[table BorrowMaterial]",
  "[table BorrowReturnMaterial]",
  "[table CollectingEvent]",
  "[table CollectingEventAttr]",
  "[table CollectingEventAttribute]",
  "[table CollectingEventAuthorization]",
  "[table CollectingTrip]",
  "[table CollectingTripAttribute]",
  "[table CollectingTripAuthorization]",
  "[table Collection]",
  "[table CollectionObjectAttr]",
  "[table CollectionObjectAttribute]",
  "[table CollectionObjectCitation]",
  "[table CollectionObjectProperty]",
  "[table CollectionRelType]",
  "[table CollectionRelationship]",
  "[table Collector]",
  "[table CommonNameTx]",
  "[table CommonNameTxCitation]",
  "[table ConservDescription]",
  "[table ConservEvent]",
  "[table Container]",
  "[table DNAPrimer]",
  "[table DNASequence]",
  "[table DNASequencingRun]",
  "[table DNASequencingRunCitation]",
  "[table Deaccession]",
  "[table DeaccessionAgent]",
  "[table Determination]",
  "[table DeterminationCitation]",
  "[table Determiner]",
  "[table Discipline]",
  "[table Disposal]",
  "[table DisposalAgent]",
  "[table DisposalPreparation]",
  "[table Division]",
  "[table ExchangeIn]",
  "[table ExchangeInPrep]",
  "[table ExchangeOut]",
  "[table ExchangeOutPrep]",
  "[table Exsiccata]",
  "[table ExsiccataItem]",
  "[table Extractor]",
  "[table FieldNotebook]",
  "[table FieldNotebookPage]",
  "[table FieldNotebookPageSet]",
  "[table FundingAgent]",
  "[table GeoCoordDetail]",
  "[table Geography]",
  "[table GeographyTreeDef]",
  "[table GeographyTreeDefItem]",
  "[table GeologicTimePeriodTreeDef]",
  "[table GeologicTimePeriodTreeDefItem]",
  "[table Gift]",
  "[table GiftAgent]",
  "[table GiftPreparation]",
  "[table GroupPerson]",
  "[table InfoRequest]",
  "[table Institution]",
  "[table InstitutionNetwork]",
  "[table Journal]",
  "[table LatLonPolygon]",
  "[table LatLonPolygonPnt]",
  "[table LithoStratTreeDef]",
  "[table LithoStratTreeDefItem]",
  "[table Loan]",
  "[table LoanAgent]",
  "[table LoanPreparation]",
  "[table LoanReturnPreparation]",
  "[table Locality]",
  "[table LocalityCitation]",
  "[table LocalityDetail]",
  "[table LocalityNameAlias]",
  "[table MaterialSample]",
  "[table OtherIdentifier]",
  "[table PcrPerson]",
  "[table Permit]",
  "[table PrepType]",
  "[table Preparation]",
  "[table PreparationAttr]",
  "[table PreparationAttribute]",
  "[table PreparationProperty]",
  "[table Project]",
  "[table ReferenceWork]",
  "[table RepositoryAgreement]",
  "[table Shipment]",
  "[table Storage]",
  "[table StorageTreeDef]",
  "[table StorageTreeDefItem]",
  "[table Taxon]",
  "[table TaxonAttribute]",
  "[table TaxonCitation]",
  "[table TaxonTreeDef]",
  "[table TaxonTreeDefItem]",
  "[table TreatmentEvent]",
  "[table VoucherRelationship]",
]
`;
